### PR TITLE
Typing: Support New Mypy Semantic Analyzer

### LINKED ISSUE #27018: Support MyPy New Semantic Analyzer
MyPy recently released a new semantic analyzer in v0.710:

https://mypy-lang.blogspot.com/2019/06/mypy-0710-released-new-semantic-analyzer.html

This will likely be the default in the next mypy release so would be good to get a head start on support. This currently causes a few errors in our code base:

```sh
$ mypy --new-semantic-analyzer pandas
pandas/core/arrays/interval.py:574: error: Name 'dtype' already defined on line 129
pandas/core/indexes/timedeltas.py:308: error: Name '_box_func' already defined on line 173
pandas/io/json/__init__.py:5: error: Name 'json' is not defined
pandas/io/json/__init__.py:5: error: Name 'normalize' is not defined
pandas/io/json/__init__.py:5: error: Name 'table_schema' is not defined
```

### PR DESCRIPTION
**Pull request by alimcmaster1**
- [x] closes #27018

### CODE REVIEW

**Comment by WillAyd** [comment]
Does this cause Mypy failures? Wasn’t think we’d add this as a setting as much as just fix the errors that using it would cause

**Comment by alimcmaster1** [comment]
^ I'm just checking that it does cause failures. This way we can verify i've actually fixed the issues

**Comment by WillAyd** [comment]
Gotcha. Just FYI you can also run `mypy --new-semantic-analyzer pandas` locally to see

**Comment by alimcmaster1** [comment]
first two errors removed, unsure of how to deal with the mypy error
```
pandas/io/json/__init__.py:5: error: Name 'json' is not defined
```

Due to the fact we have: ( in our __init__.py)
```
del json, normalize, table_schema  # noqa
```

Any advice @WillAyd ?

**Comment by WillAyd** [comment]
Does converting the imports to abolute help at all? Seems like kind of a strange __init__ file - maybe modeling it after pandas.io.excel would help and allow for removal of noqa

File: pandas/io/json/__init__.py [hunk 1]
```diff
@@ -1,5 +1,3 @@
-from .json import dumps, loads, read_json, to_json  # noqa
-from .normalize import json_normalize  # noqa
-from .table_schema import build_table_schema  # noqa
-
-del json, normalize, table_schema  # noqa
+from pandas.io.json._json import dumps, loads, read_json, to_json
+from pandas.io.json._normalize import json_normalize
+from pandas.io.json._table_schema import build_table_schema
```
Comment (WillAyd) [inline pandas/io/json/__init__.py]: Think you need to define __all__

File: pandas/io/json/_json.py [hunk 2]
```diff
@@ -19,8 +19,8 @@
 from pandas.io.formats.printing import pprint_thing
 from pandas.io.parsers import _validate_integer
 
-from .normalize import _convert_to_line_delimits
-from .table_schema import build_table_schema, parse_table_schema
+from ._normalize import _convert_to_line_delimits
```
Comment (WillAyd) [inline pandas/io/json/_json.py]: Not sure about renaming these modules as part of this (though maybe OK) - was this causing issues or are you just trying to make them more explicitly private?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/json/_json.py (hunk 2 above)
Comment (alimcmaster1) [reply pandas/io/json/_json.py]: Trying to make them more explicitly private and match convention in io/excel and io/msgpack. Since we are not calling del any more in the __init__ file I thought this made sense

File: pandas/io/json/__init__.py (hunk 1 above)
Comment (alimcmaster1) [reply pandas/io/json/__init__.py]: Ahh yep added!

File: pandas/io/json/_json.py (hunk 2 above)
Comment (WillAyd) [reply pandas/io/json/_json.py]: Sure - might be fine just thinking through impacts but leave as is for now. I think if you define `__all__` in the init should take care of mypy errors

**Comment by WillAyd** [comment]
Just remove the noqa - shouldn't need them

**Comment by alimcmaster1** [comment]
```
pandas\tests\indexes\multi\test_format.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <contextlib._GeneratorContextManager object at 0x000002299F625860>
type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):
        if type is None:
            try:
>               next(self.gen)
E               AssertionError: Did not see expected warning of class 'FutureWarning'.
```

**Comment by WillAyd** [comment]
Code changes look good. Can you remove changes to mypy.ini and any requirements files?

@jreback any objection to renaming of files in json folder?

**Comment by jreback** [comment]
no objections

**Comment by alimcmaster1** [comment]
Thanks for the review @WillAyd . 

Just for my understanding what’s your motivation for removing the changes to mypy.ini it’s whats validating the mypy semantic analysis in CI - to avoid anyone introducing further errors?

Thanks

**Comment by WillAyd** [comment]
> Just for my understanding what’s your motivation for removing the changes to mypy.ini it’s whats validating the mypy semantic analysis in CI - to avoid anyone introducing further errors?

Because we don't need to really pin a dependency here. Any of these changes are backwards compatible we are just preventing future mypy versions from breaking; no need to force a min version

**Comment by alimcmaster1** [comment]
^ ignore me trying to fix up this PR on mobile went badly.. Will update tomorrow.

File: pandas/io/json/__init__.py [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
 
-del json, normalize, table_schema  # noqa
+__all__ = [
+    dumps,
+    loads,
+    read_json,
+    to_json,
+    json_normalize,
+    build_table_schema
```
Comment (jorisvandenbossche) [inline pandas/io/json/__init__.py]: those should be strings I think?

**Review by WillAyd — APPROVED** [review]
lgtm - over to you @jorisvandenbossche

**Comment by jreback** [comment]
@alimcmaster1 isn't the commit adding the semantic analyzer removed? IOW this doesn't look like it is triggering it.

File: pandas/core/arrays/interval.py [hunk 4]
```diff
@@ -139,7 +139,6 @@
     )
 )
 class IntervalArray(IntervalMixin, ExtensionArray):
-    dtype = IntervalDtype()
```
Comment (jreback) [inline pandas/core/arrays/interval.py]: how is this working? do we already have this defined elsewhere?

File: pandas/core/indexes/timedeltas.py [hunk 5]
```diff
@@ -176,7 +176,6 @@ def _join_i8_wrapper(joinf, **kwargs):
 
     _freq = None
 
-    _box_func = TimedeltaArray._box_func
```
Comment (jreback) [inline pandas/core/indexes/timedeltas.py]: why is this removed?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
I asked to remove the semantic analyzer from the config because it naturally becomes the default in the next Mypy release. Figured no need to set then and not setting maintains better backwards compat

**Comment by jreback** [comment]
> I asked to remove the semantic analyzer from the config because it naturally becomes the default in the next Mypy release. Figured no need to set then and not setting maintains better backwards compat

ok, then we should set a min on mypy (when this comes out)? or at least on our CI do that?

**Comment by WillAyd** [comment]
We don’t have to. The new analyzer is more strict so anything done here works with or without it. 

But yea if we wanted to play it safe could set min version to next release when it comes out

File: pandas/io/json/__init__.py (hunk 3 above)
Comment (alimcmaster1) [reply pandas/io/json/__init__.py]: Updated

File: pandas/core/arrays/interval.py (hunk 4 above)
Comment (alimcmaster1) [reply pandas/core/arrays/interval.py]: Exactly already defined as a class property

File: pandas/core/indexes/timedeltas.py (hunk 5 above)
Comment (alimcmaster1) [reply pandas/core/indexes/timedeltas.py]: Same issue again already defined as a class property

**Review by jreback — APPROVED** [review]

File: pandas/io/json/_json.py (hunk 2 above)
Comment (jreback) [reply pandas/io/json/_json.py]: do we actually need to rename modules for this to pass?

Comment (alimcmaster1) [reply pandas/io/json/_json.py]: Previously in the  __init__ file we were deleting the module reference - which the analyser didn’t like. We instead renamed the modules with a _ to make it obvious they are private. ( this seemed to match the convention in the io/excel package)

Comment (jreback) [reply pandas/io/json/_json.py]: ok, can you de-privatize convert_to_line_delimits then (as now the module itself is private)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by alimcmaster1** [comment]
Updated as per your comments and green @jreback

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @alimcmaster1
